tools/crm_resource_runtime.c

```diff
@@ -1713,3 +1752,87 @@ cli_resource_move(const char *rsc_id, const char *host_name, cib_t * cib, pe_wor
 
     return rc;
 }
+
+static void
+cli_resource_why_without_rsc_and_host(cib_t *cib_conn,GListPtr resources)
+{
+    GListPtr lpc = NULL;
+    GListPtr hosts = NULL;
+
+    for (lpc = resources; lpc != NULL; lpc = lpc->next) {
+        resource_t *rsc = (resource_t *) lpc->data;
+        rsc->fns->location(rsc, &hosts, TRUE);
+	if ( hosts == NULL ) {
+	    printf("Resource %s is not running",rsc->id);
+            cli_resource_check(cib_conn, rsc);
+	} else {
+	    printf("Resource %s is running",rsc->id);
+	}
+     }
+
+     g_list_free(hosts);
+     hosts = NULL;
+}
+
+static void
+cli_resource_why_with_rsc_and_host(cib_t *cib_conn,GListPtr resources,const char* rsc_id,const char* host_uname)
+{
+    resource_t *rsc = NULL;
+
+    rsc = pe_find_resource(resources, rsc_id);
+    if(!(resource_is_running_on(rsc,host_uname))) {
+        cli_resource_check(cib_conn, rsc);
+    } else {
+        printf("Resource %s is not running on host %s",rsc->id,host_uname);
+    }
+}
+
+static void 
+cli_resource_why_without_rsc_with_host(cib_t *cib_conn,GListPtr resources,const char* host_uname)
+{
+    GListPtr allResources = get_all_resources(host_uname,resources);
```

Comment:
I think we can avoid creating a new function to get all resources. You can find the node in the data set using pe_find_node() (see the 'U' handling for an example), and pass the node entry rather than the host name to cli_resource_why(). There is a node->details->allocated_rsc member that is a list of all resources assigned to that node (whether or not they are running). There is also node->details->running_rsc with just the ones that are running, so you can use subtract_lists() to get the inactive ones. The node name is available as node->details->uname.